flycheck.el

```diff
@@ -5625,129 +5625,172 @@ about TSLint."
+
+    ;; Turn each span into a flycheck error
+    (dolist (span spans)
+      (let-alist span
+        ;; Children lack any filename/line/column information, so we use
+        ;; those from the primary span
+        (when .is_primary
+          (setq primary-filename .file_name
+                primary-line .line_start
+                primary-column .column_start))
+        (push
+         (flycheck-error-new-at
+          .line_start
+          .column_start
+          ;; Non-primary spans are used for notes
+          (if .is_primary error-level 'info)
+          (if .is_primary
+              ;; Primary spans may have labels with additional information
+              (concat error-message (when .label
+                                      (format " (%s)" .label)))
+            ;; If the label is empty, fallback on the error message,
+            ;; otherwise we won't be able to display anything
+            (or .label error-message))
+          :id error-code
+          :checker checker
+          :buffer buffer
+          :filename .file_name)
+         errors)))
+
+    ;; Then we turn children messages into flycheck errors pointing to the
+    ;; location of the primary span.  According to the format, children
+    ;; may contain spans, but they do not seem to use them in practice.
+    (dolist (child children)
+      (let-alist child
+        (push
+         (flycheck-error-new-at
+          primary-line
+          primary-column
+          'info
+          .message
+          :id error-code
+          :checker checker
+          :buffer buffer
+          :filename primary-filename)
+         errors)))
+
+    ;; If there are no spans, the error is not associated with a specific
+    ;; file but with the project as a whole.  We still need to report it to
+    ;; the user by emitting a corresponding flycheck-error object.
+    (unless spans
+      (push (flycheck-error-new-at
+             ;; We have no specific position to attach the error to, so
+             ;; let's use the top of the file.
+             1 1
+             error-level
+             error-message
+             :id error-code
+             :checker checker
+             :buffer buffer)
+            errors))
     (nreverse errors)))
 
+(defun flycheck-parse-rust-filter-json (output)
+  "Return parsed JSON data from OUTPUT generated by Cargo or rustc.
+
+When JSON output is requested, Cargo and rustc both emit a mix of
+plain text lines and JSON lines.  This function ignores the plain
+text lines, and parses the JSON lines before returning them in a
+list."
+  (let ((json-array-type 'list)
+        (json-false nil)
+        ;; Skip the plain text lines in OUTPUT, keep the JSON lines.
+        (json-lines (seq-filter (lambda (line)
+                                  (string-match-p "^{" line))
+                                (split-string output "\n"))))
+    ;; Each JSON line is a JSON object.
+    (seq-map #'json-read-from-string json-lines)))
```

Comment:
In the interest of science, I conducted a micro-benchmark using `benchmark-run-compiled`.  On a large buffer (300 JSON error objects mixed with plain text), running your version takes 0.05 seconds, while the ~seq-map~ version takes 0.06 seconds.  Going to 100 iterations shows the same difference (5.34 vs 6.9), with slightly more variance on the ~seq-map~ version few runs.

I don't think the performance difference would matter in practice.  But since I do like your version more anyway – very raw elisp, while still being readable – I'll go with it :)